crates/ra_hir_expand/src/lib.rs

```diff
@@ -301,3 +301,40 @@ impl<T> InFile<T> {
         db.parse_or_expand(self.file_id).expect("source created from invalid file")
     }
 }
+
+impl<T: Clone> InFile<&T> {
+    pub fn cloned(&self) -> InFile<T> {
+        self.with_value(self.value.clone())
+    }
+}
+
+impl InFile<SyntaxNode> {
+    pub fn ancestors_with_macros<'a>(
+        self,
+        db: &'a impl crate::db::AstDatabase,
+    ) -> impl Iterator<Item = InFile<SyntaxNode>> + 'a {
+        SyntaxNodeAncestorsWithMacros { db, node: Some(self) }
```

Comment:
Hm, seems like `std::iter::successors` would also work here?

Reply:
Yeah, actually..

Reply:
Btw, this feels very similar to `.contexts` in IntelliJ Rust: https://github.com/intellij-rust/intellij-rust/blob/39daa3ea0df34255a434bdb24f27fed41f4d2df7/src/main/kotlin/org/rust/lang/core/psi/ext/PsiElement.kt#L31-L34